ext/url/01_urlpattern.js

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+// @ts-check
+/// <reference path="../../core/internal.d.ts" />
+/// <reference path="../../core/lib.deno_core.d.ts" />
+/// <reference path="../webidl/internal.d.ts" />
+/// <reference path="./internal.d.ts" />
+/// <reference path="./lib.deno_url.d.ts" />
+
+"use strict";
+
+((window) => {
+  const core = window.Deno.core;
+  const webidl = window.__bootstrap.webidl;
+  const {
+    ArrayPrototypeMap,
+    ObjectKeys,
+    ObjectFromEntries,
+    RegExp,
+    RegExpPrototypeExec,
+    RegExpPrototypeTest,
+    Symbol,
+    TypeError,
+  } = window.__bootstrap.primordials;
+
+  const _components = Symbol("components");
+
+  /**
+   * @typedef Components
+   * @property {Component} protocol
+   * @property {Component} username
+   * @property {Component} password
+   * @property {Component} hostname
+   * @property {Component} port
+   * @property {Component} pathname
+   * @property {Component} search
+   * @property {Component} hash
+   */
+
+  /**
+   * @typedef Component
+   * @property {string} patternString
+   * @property {RegExp} regexp
+   * @property {string[]} groupNameList
+   */
+
+  class URLPattern {
+    /** @type {Components} */
+    [_components];
```

Comment:
wonder why not use js private field?

Reply:
Because private fields can not be inspected internally. See for example how the `URLSearchParams` interface needs to access the internal parameters of `URL`.

All our internal code uses symbols for internal properties.